add linting via dscanner

wilzbach opened the pull request:
I think we should try putting the linting off to a real bot.
This PR adds a execution of [dscanner](https://github.com/Hackerpilot/Dscanner).

We probably want to play with this for a while, maybe disable some checks or fix the codebase until we set the `--enforce` flag, which will exit the CI with an error.
Currently it outputs many messages, see e.g. [here](https://travis-ci.org/wilzbach/mir/jobs/122970079)

Other
- It is only run on travis subbuild - hence the version check (I reused it for the s3 check)
- As there is no binary release of dscanner, I created `releases.dlang.io` (a s3 bucket to which I will upload such release) - this speeds up the build! Otherwise there is still dub fetch & dub run.

9il commented:
I don't think dscanner is goof enought

wilzbach commented:
> I don't think dscanner is good enought

AFAIK it's the best tool out there and we can just disable the warnings that we don't like.
In my opinion a linting is better than no linting?

9il commented:
ping me when you would think PR is ready

wilzbach commented:
@9il I think I am getting somewhere, but for the remaining issues I would like to hear your opinion and help. I do consider static analysis of code important as it can prevent a couple of errors and it ensures that the codes __sticks to a style guide__ - the latter being the main reason for my attempt here.
Dscanner checks for a lot of stuff - have a look at the `.dscanner.ini`. It was really helpful to realize that the new combinatorics module didn't document itself properly.

You can run it locally with:

```
dub fetch dscanner --version=~master
dub run dscanner -- --config .dscanner.ini --styleCheck source
```
Or install the `dscanner-git` package on Arch Linux. I recommend storing the binary somewhere as you avoid the `dub` checking and can pass the spammy error log to dev/null. I use

```
dscanner --config .dscanner.ini --styleCheck source 2> /dev/null
```

I grouped them after their warning types, so that it's easier to look at them.
Please do consider that disabling a check because of one or two edge cases results in we no benefit from this help in the future. I still didn't set the `--enforce` flag, but I plan to do so after we resolved remaining the conflicts below.

Naming style guide
-------------------------

From Phobos: functions + variables in camelCase.

```
source/mir/las/sum.d(2125:11)[warn]: Function name 'SummationAlgo' does not match style guidelines.
source/mir/las/sum.d(2135:11)[warn]: Function name 'SummationAlgo' does not match style guidelines.
source/mir/ndslice/internal.d(174:21)[warn]: Variable name 'DynamicArrayDimensionsCount' does not match style guidelines.
source/mir/ndslice/internal.d(176:21)[warn]: Variable name 'DynamicArrayDimensionsCount' does not match style guidelines.
source/mir/ndslice/internal.d(256:8)[warn]: Struct name '_Slice' does not match style guidelines.
source/mir/ndslice/slice.d(1042:21)[warn]: Variable name 'PureN' does not match style guidelines.
```

Subtracting from `.length`
-----------------------------------

```
source/mir/ndslice/selection.d(221:29)[warn]: Avoid subtracting from '.length' as it may be unsigned.
source/mir/ndslice/selection.d(222:29)[warn]: Avoid subtracting from '.length' as it may be unsigned.
```

Zero-parameter '@property' function
-------------------------------------------------

```
source/mir/ndslice/selection.d(937:22)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
source/mir/ndslice/selection.d(979:22)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
source/mir/ndslice/selection.d(1422:22)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
source/mir/ndslice/slice.d(1327:14)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
source/mir/ndslice/slice.d(1374:14)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
source/mir/ndslice/slice.d(2845:14)[warn]: Zero-parameter '@property' function should be marked 'const', 'inout', or 'immutable'.
```

'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' should be const
-----------------------------------------------------------------------------------------------

```
source/mir/ndslice/slice.d(1591:10)[warn]: Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.
source/mir/ndslice/slice.d(1609:10)[warn]: Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.
```

has method 'opEquals', but not 'toHash'
-----------------------------------------------------

```
source/mir/ndslice/slice.d(1027:8)[warn]: 'Slice' has method 'opEquals', but not 'toHash'.
```

Avoid use of the comma expression
-----------------------------------------------

```
source/mir/ndslice/selection.d(763:66)[warn]: Avoid using the comma expression.
```


Disabled flags
-------------------

Moreover I disabled a couple of flags temporarily - I do plan to fix the codebase at a later point and toggle this on, but let's go step by step.

Here's the list with the reason:

- undocumented public documentation: it correctly realizes that for `///` the documentation is undefined. not sure yet whether this is a bug. Should we write `/// ditto`?
- "Local imports should specify the symbols being imported to avoid hiding local symbols": there were too many violations.
- unused variable: issue [#326](https://github.com/Hackerpilot/Dscanner/issues/326)
- could be immutable: issue [#325](https://github.com/Hackerpilot/Dscanner/issues/325)
- checks for labels with the same name as variables: issue [#323](https://github.com/Hackerpilot/Dscanner/issues/323)

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -167,6 +166,24 @@ unittest
+
+/// ditto
 struct IndexedRoR(Collection, Range)
 if (isInputRange!Range)
```
this didn't have docs ...

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -308,42 +274,13 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
-
+/// ditto
 struct Permutations
 {
```
docs were missing for all the structs - the order is now

`method`
`makeMethod` /// ditto
struct `Method`  /// ditto
unittest
dispose(Method ..)

wilzbach commented on source/mir/las/sum.d:
```diff
@@ -374,9 +374,13 @@ struct Summator(T, Summation summation)
+        Fast pairwise is enabled for floating point, complex numbers.
+        If SimD is enabled it also works for other numeric types.
+        */
+        enum bool fastPairwise =
```
public method - requires doc. can u check whether this makes sense?

9il replied:
should be private

wilzbach commented on source/mir/las/sum.d:
```diff
@@ -374,9 +374,13 @@ struct Summator(T, Summation summation)
+        */
+        enum bool fastPairwise =
+            isFloatingPoint!F ||
+            isComplex!F ||
```
trailing whitespace ...

wilzbach commented on source/mir/las/sum.d:
```diff
@@ -1822,16 +1828,20 @@ nothrow unittest
+            assert(summator.isInfinity() == r.isInfinity()
+                 || (r == -double.max && s == -double.infinity)
+                 || (r == double.max && s == double.infinity));
+            assert((nextDown(s) <= r && r <= nextUp(s)) || (s.isNaN && r.isNaN));
```
dscanner checks for such tricky cases where the prevalence of `&&` over `||` matters.

9il replied:
turn it off please

wilzbach replied:
Are u sure? it does make the code better to read. The idea is to avoid confusing expressions:

```
if (a && b || c) // bad
if (a && (b || c)) // good
```

It checks is if there are multiple `||` and `&&` in one expression,
so `1 && 2 && 3` will not trigger it.

It is called ["logical precedence check"](https://github.com/Hackerpilot/Dscanner/blob/master/src/analysis/logic_precedence.d)

9il replied:
no, i don't like braces, expressions become less readable

9il replied:
> Are u sure?

Yes

wilzbach replied:
> no, i don't like braces, expressions become less readable

I disagree, but you are the "boss" ;-)
Removed.

wilzbach commented on source/mir/las/sum.d:
```diff
@@ -2204,7 +2214,9 @@ private template SummationType(F)
-            pragma(msg, "Note: Summation algorithms on x86 use 80bit representation for single and double floating point numbers.");
+            enum note =  "Note: Summation algorithms on x86 use 80bit representation"
+                       ~ "for single and double floating point numbers.";
+            pragma(msg, note);
```
over 120 chars

wilzbach commented on source/mir/ndslice/iteration.d:
```diff
@@ -127,12 +127,10 @@ template swapped(size_t dimensionA, size_t dimensionB)
     auto swapped(size_t N, Range)(Slice!(N, Range) slice)
     {
         {
-            enum i = 0;
```
unused variable ...

9il commented:
> Naming style guide

turn off please

wilzbach commented on source/mir/ndslice/selection.d:
```diff
@@ -917,7 +917,7 @@ auto byElement(size_t N, Range)(auto ref Slice!(N, Range) slice)
 
             static if (canSave!PureRange)
-            auto save() @property
+            auto save()
```
`@property` methods with zero arguments should return `const`.
Changing this as it still can be used via UCFS

wilzbach commented on source/mir/ndslice/selection.d:
```diff
@@ -1686,12 +1686,12 @@ template IotaSlice(size_t N)
 // undocumented
 // zero cost variant of `std.range.iota`
-struct IotaMap()
+private struct IotaMap()
```
public method without documentation should be set to private or protected

9il commented:
I would not merge this until the Phobos merged or you remove changes in `ndslice` package

9il commented:
Please move all changes to separate PR (and without ndslice)

wilzbach commented:
@9il I tried to describe why I did some of the changes - hope this helps to understand ;-)
Will separate now.

wilzbach commented:
> Please move all changes to separate PR (and without ndslice)

I rebased this one and moved the ndslice changes to #89 which you can keep open.

9il commented:
split any changes from linter addition

wilzbach commented:
> split any changes from linter addition

done.

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -265,39 +264,6 @@ if (isInputRange!Range && hasLength!Range)
-    assert(a.front.equal([0, 1]));
-}
-
 /// ditto
```
should have separate comment

wilzbach replied:
was this note for `Permutations` - imho `makePermutations` should be bundled with `permutations`

9il replied:
hmm. ok

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -308,42 +274,13 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
-}
-
+/// ditto
 struct Permutations
```
should have separate comment

wilzbach replied:
"Forward range of Permutations?"

Or more detailed? Most of the stuff is already explained at the `permutations` method, hence the `ditto`.

9il replied:
> Or more detailed

A little bit more and links to make**\* and ***

9il replied:
See_also: links

wilzbach replied:
already in the working - how can i link to the according `dispose`?

I know only `$(LREF dispose)`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -308,42 +274,13 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
     private bool _empty;
 
-    this(uint[] state, uint[] indices) @safe pure nothrow @nogc
+    private this(uint[] state, uint[] indices) @safe pure nothrow @nogc
```
make it public please with `///`

wilzbach replied:
But then we should explain that those arrays just need to be empty?

9il replied:
no, just need to have proper size

9il commented:
BTW, i think documentation is more important now because we included combinatorics and will include sparse soon.

9il commented:
only ndslice has docs on dlang.org

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -410,6 +366,70 @@ struct Permutations
+}
+
+///
+static if (__VERSION__ > 2069) @nogc unittest
```
I think as long as allocators are still new we should make the unittest public.

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -308,41 +277,28 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
 
-    import std.experimental.allocator.mallocator;
+It will always generate the permutations from 0 to `n - 1`,
+use $(LREF indexedRoR) to map it to your range.
```
how do you like this sentence - is it clear?

9il replied:
my english is not well, but it seems that present simple is better then future

wilzbach replied:
> my english is not well,

no worries - your coding is well!
Removed the `will`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -308,41 +277,28 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+en.wikipedia.org/wiki/Heap%27s_algorithm, Heap's algorithm).
 
-    import std.experimental.allocator.mallocator;
+It always generate the permutations from 0 to `n - 1`,
```
generateS?

wilzbach replied:
refresh :)

wilzbach commented:
> BTW, i think documentation is more important now because we included combinatorics and will include sparse soon.

We will have a look during the weekend ;-)

wilzbach commented:
It seems like I "force-pushed" travis to its limits, but now it's passing finally :)

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -167,6 +166,24 @@ unittest
+onto a custom `range`.
+
+Params:
+    collenction = Range to be projected from
```
should be lower case

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -308,41 +277,28 @@ Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+the total number of elements is `n^k`.
 
+See_Also:
+    $(LREF cartesianPower), $(LREF makeCartesianPower)
```
Copy past error

wilzbach replied:
already fixed - note at the wrong place?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -251,6 +250,9 @@ Params:
     Forward range, which yields the permutations
+
+See_Also:
+    $(LREF dispose)
```
remove $(LREF dispose) and add $(LREF Permutations)

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -251,6 +300,9 @@ Params:
     Forward range, which yields the permutations
+
+See_Also:
+    $(LREF Permutations), $(LREF dispose)
```
As it is bundled with `makePermutations`, linking to `dispose` could be useful to the user

9il replied:
`dispose` make refer to another `dispose`

wilzbach replied:
> dispose make refer to another dispose

Ehm sry - what do you mean?

9il replied:
you have multiple dispose declarations

wilzbach replied:
So there is no way to link to the "correct" one?

9il replied:
no way

wilzbach replied:
I just tested and if I click on `dispose` it will get to such a page - is this good enough or should we trust that the unittest does the explanation?

![image](https://cloud.githubusercontent.com/assets/4370550/14567575/5b2cef92-033d-11e6-8fba-96ccd58c8996.png)

9il replied:
this is not good enough  

> On 15 апр. 2016 г., at 18:08, Sebastian Wilzbach notifications@github.com wrote:
> 
> In source/mir/combinatorics/package.d https://github.com/DlangScience/mir/pull/88#discussion_r59898839:
> 
> > @@ -251,6 +300,9 @@ Params:
> > 
> >  Returns:
> >      Forward range, which yields the permutations
> > +
> > +See_Also:
> > -    $(LREF Permutations), $(LREF dispose)
> >   I just tested and if I click on dispose it will get to such a page - is this good enough or should we trust that the unittest does the explanation?
> 
>  https://cloud.githubusercontent.com/assets/4370550/14567575/5b2cef92-033d-11e6-8fba-96ccd58c8996.png
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub https://github.com/DlangScience/mir/pull/88/files/21c97a5060d40c21e7e415ee78bab94c461e591f#r59898839

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -265,41 +317,8 @@ if (isInputRange!Range && hasLength!Range)
-
 /// ditto
-Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+Permutations makePermutations(Allocator)(size_t n, auto ref Allocator alloc)
```
had the wrong order ;-)

9il replied:
No, allocator should be always first for make\* functions

> On 15 апр. 2016 г., at 17:38, Sebastian Wilzbach notifications@github.com wrote:

wilzbach replied:
Ok will change the other functions

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -359,6 +366,10 @@ struct Permutations
+
+        // factorial
+        foreach (i; 1..indices.length + 1)
+            _max_states *= i;
```
all the other ranges provide a `length` function - i don't think it makes a noticeable overhead. Alternatively we could also compute it on-demand only

wilzbach commented:
Hmm it seems that there were still a lot of stuff to do for the combinatorics module - here's a list of what i changed:

- allocator order for makeCombinations,Repeat,CartesianPower
- length for permutations (all other collections have this too)
- unittests for length
- unittest for invalid input
- unittest for indexed
- added const to `@property` attributes like `bool` and `length`
- reworked the description for `IndexedRoR

9il commented:
> allocator order for makeCombinations,Repeat,CartesianPower

allocators should be first argumensts

wilzbach commented:
> allocators should be first argumensts

Yep those had the allocater as last argument

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -613,14 +718,59 @@ pure nothrow @safe unittest
 the number of combinations is `binomial(n, k)`.
 
-For example $(D"AB".combinations(2).array) returns $(D["AB"]).
+For example `"AB".combinations(2).array` returns `["AB"]`.
```
should I remove `.array` here?

9il replied:
remove this example from the docs. And add constraints, that templates are note accepts narrow strings.

wilzbach commented:
@9il CI was failing due to an added tests which included the `Mallocator` - should be fixed now. Any other objections? :)

wilzbach commented:
> remove this example from the docs. And add constraints, that templates are note accepts narrow strings.

I created a module level unittest as an overview over the iterators.
Moreover I realized that `indexed` from std.range actually needs a random access range and thus corrected all template constraints. A narrow string can't be a random access string :)

wilzbach commented:
(rebased to latest master - as we are moving quite fast tonight)

wilzbach commented:
> Failed to download http://code.dlang.org/packages/libevent/2.0.1+2.0.16.zip: 404 Not Found

why? :(

9il commented:
> why? :(

Just remove ddox. I don't like vibe.d and ddox ;/

We can use https://github.com/kiith-sa/harbored-mod, it is the best but needs small update. Probably we would need to add macros file from dlang

wilzbach commented:
> We can use https://github.com/kiith-sa/harbored-mod, it is the best but needs small update.

Okay will send a PR later ;-)

wilzbach commented:
> Okay will send a PR later ;-)

Hmm seems that we have to fix libdparse first ...
https://github.com/kiith-sa/harbored-mod/issues/62
